### PR TITLE
Support `log_every_n_steps` with validate|test

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lightning/issues/10436

Logging per step during fit's validation, regular validation, testing, or predicting is not generally useful when the logged values depend on the optimization process.
However, sometimes you want to log values that do not depend on the optimization process. Examples of this are throughput-related metrics (as in #18848) or batch-specific metrics.

This PR adds support for tuning the logging interval under these circumstances.

This is only a breaking change if the user was calling `self.log(..., on_step=True)`.
The previous behavior is equal to setting `log_every_n_steps=1` now. Before, this value was ignored.

----
:books: Documentation preview :books:: https://pytorch-lightning--18895.org.readthedocs.build/en/18895/

cc @borda @carmocca @justusschock

### CODE REVIEW

File: tests/tests_pytorch/trainer/logging_/test_eval_loop_logging.py [hunk 1]
```diff
@@ -548,11 +548,11 @@ def test_step(self, batch, batch_idx):
         call(metrics={"train_loss": ANY, "epoch": 0}, step=0),
         call(metrics={"valid_loss_0_step": ANY, "valid_loss_2": ANY}, step=0),
         call(metrics={"valid_loss_0_step": ANY, "valid_loss_2": ANY}, step=1),
-        call(metrics={"valid_loss_0_epoch": ANY, "valid_loss_1": ANY, "epoch": 0}, step=0),
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/logging_/test_eval_loop_logging.py:551]: The step value here corresponded to the training step. Now it's for the validation step

**Comment by awaelchli** [comment]
Hey @carmocca 
There are some significant differences here that should be discussed. Here is my simple example:

```py
import os

import torch

from lightning import seed_everything
from lightning.pytorch import LightningModule, Trainer
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch.loggers import TensorBoardLogger


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    seed_everything(0)
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    logger = TensorBoardLogger("tb_logs", name="my_model")

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        logger=logger,
        limit_train_batches=10,
        limit_val_batches=10,
        limit_test_batches=10,
        num_sanity_val_steps=4,
        max_epochs=3,
        enable_model_summary=False,
        log_every_n_steps=2,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.validate(model, val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()

```

Running `tensorboard --logdir tb_logs` and comparing master with this branch.

Red: master
Blue: this PR

**Issue 1:**

image

Here we see that the new behavior no longer logs the epoch value at the epoch-end step (blue stops at step 29). It falling short now means there would be a gap when resuming:
image

**Issue 2:**

image

Here the test loss no longer gets logged at the global step, but the step is local to the test stage. 

**Issue 3:**
image

Finally, a few things are going on for the validation. 1) The fit-validation values are no longer logged, because they don't fall into the logging interval (they are epoch-end). 2) Without using the global step for logging fit-validation logging, you can no longer track the validation loss across your training. It all gets logged to the same step value.

**Comment by carmocca** [comment]
I'll look at Issue 1 and 3. Thanks.
Issue 2 is expected and the point of this PR, that is, testing using its own index instead of the training global step.
